### PR TITLE
NO-1963: Add UI and unit tests for navigation (goto) and focus callbacks

### PR DESCRIPTION
**Pull request by vivekjyani**
### 1. Context / Spec
- **Ticket/Spec:** NO-1963
- **Summary:** Add UI and unit test coverage for the navigation (goto) feature: field/page focus, onFocus/onBlur when using `goto(path, GotoConfig(open:, focus:))`, and collection/table focus. Ensures navigation behavior is regression-tested.

---

### 2. What Changed

| Area | Files / changes |
|------|-----------------|
| **Unit tests** | `JoyfillSwiftUIExample/JoyfillTests/NavigationJSONTests.swift` (new) – navigation/goto scenarios using Navigation.json |
| **UI tests – focus** | `JoyfillSwiftUIExample/JoyfillUITests/FocusCallbackUITests/FocusCallbackUITests.swift` (new) – onFocus firing for text, number, dropdown, table row/column, collection row/column when using goto with focus |
| **UI tests – changelogs** | `JoyfillSwiftUIExample/JoyfillUITests/ChangeUITests/OnChangeHandlerUITests.swift` – page focus changelog tests by navigator |
| **UI tests – collection** | `JoyfillSwiftUIExample/JoyfillUITests/Collection/CollectionFieldTests.swift` – collection focus test |
| **Test infra** | `JoyfillSwiftUIExample/JoyfillUITests/JoyfillUITestsBaseClass.swift` – helpers for focus/changelog assertions and goto launch args |
| **Example app** | `JoyfillSwiftUIExample/JoyfillExample/UITestFormContainerView.swift` – `triggerGotoFromLaunchArgumentsIfNeeded()` and focus/blur result handling for UI tests |
| **Project** | `JoyfillSwiftUIExample/JoyfillExample.xcodeproj/project.pbxproj`, `JoyfillExampleApp.swift` – wire new test target and launch args |

---

### 3. Implementation Decisions
- **Goto from launch args:** UI tests pass `--goto-path`, `--goto-open`, `--goto-focus` so the example app performs goto on load; avoids flaky taps and keeps tests deterministic.
- **Focus/blur assertion:** Example app exposes focus/blur events as JSON in a label (`focusBlurResultfield`); UI tests read it and assert expected field IDs and event kinds.
- **Navigation.json:** Shared between unit tests (DocumentEditor) and UI tests (example app) so behavior is consistent and one fixture covers both layers.
- **Stable identities:** Rely on existing column-based identity for table/collection refresh; tests assume IDs from Navigation.json.

---

### 4. Screenshot / Video
_Optional: add a short clip of goto + focus flow (e.g. opening a table row and seeing focus on a cell)._

---

### 5. Public API & Docs
- [ ] **Public API changed?** No
- [ ] **Docs need updating?** No
- **Notes:** N/A

---

### 6. Tests
- [x] New tests added in this PR

**Notes:** New `NavigationJSONTests` (unit) and `FocusCallbackUITests` + updates in `OnChangeHandlerUITests` and `CollectionFieldTests` (UI). Covers goto page/field/row/column, focus callback firing, and changelog behavior for navigation.

---

### 7. Notes for Reviewers
- Run UI tests with the same simulator/device; focus tests depend on `focusBlurResultfield` and goto launch args.
- `NavigationJSONTests` use `Navigation.json` from the example app bundle; no new package dependencies.
- One fix commit in the branch: “Fix build error” (likely test or project config).

### CODE REVIEW

File: JoyfillSwiftUIExample/JoyfillExample/UITestFormContainerView.swift [hunk 1]
```diff
@@ -71,20 +71,38 @@ struct UITestFormContainerView: View {
...
+    private func triggerGotoFromLaunchArgumentsIfNeeded() {
+        let args = CommandLine.arguments
+        guard let pathIndex = args.firstIndex(of: "--goto-path"),
+              pathIndex + 1 < args.count else { return }
+        let path = args[pathIndex + 1]
+        let open = args.contains("--goto-open")
+        let focus = args.contains("--goto-focus")
+        let gotoConfig = GotoConfig(open: open, focus: focus)
+        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
+            _ = self.documentEditor.goto(path, gotoConfig: gotoConfig)
```
Comment (VishnuBishnoi) [inline JoyfillSwiftUIExample/JoyfillExample/UITestFormContainerView.swift]: The fixed 0.5s delay before calling `goto` could be flaky on slower CI machines or when the form has many fields. If the form isn't fully laid out within 0.5s, `goto` may silently fail.

Consider either:
- Observing a "form did finish loading" signal from `DocumentEditor` (if available) and triggering goto in the callback.
- Or at minimum, increasing the delay and adding a comment noting it's a timing workaround:
```swift
// Delay to ensure form layout is complete before navigation.
// If flaky on CI, increase this value.
DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { ... }
```

File: JoyfillSwiftUIExample/JoyfillUITests/FocusCallbackUITests/FocusCallbackUITests.swift [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        let events = focusBlurEventsFromApp()
+        return events.contains { dict in
+            guard dict["kind"] as? String == "focus",
+                  let fieldEvent = dict["fieldEvent"] as? [String: Any],
+                  let id = fieldEvent["fieldID"] as? String else { return false }
+            return id == fieldID
+        }
+    }
+
+    func waitForFocusBlurResult(timeout: TimeInterval = 3) {
```
Comment (VishnuBishnoi) [inline JoyfillSwiftUIExample/JoyfillUITests/FocusCallbackUITests/FocusCallbackUITests.swift]: The `waitForFocusBlurResult` default timeout of 3 seconds includes the 0.5s delay in `triggerGotoFromLaunchArgumentsIfNeeded` plus form layout time. On CI this may be tight.

Also, if the goto fails silently (e.g., timing issue), `waitForFocusBlurResult` will timeout and `focusBlurEventsFromApp()` returns `[]`, making `hasFocusEventWithFieldID` return `false`. The test will fail with a generic assertion message rather than indicating *why* — was it a timing issue? Did goto return `.failure`?

Consider asserting the goto result as well (e.g., expose it via another label), or at least logging the raw `focusBlurResultfield` content in failure messages.

File: JoyfillSwiftUIExample/JoyfillTests/NavigationJSONTests.swift [hunk 3]
```diff
@@ -929,4 +929,172 @@ final class NavigationJSONTests: XCTestCase {
...
+        }
+        
+        func onCapture(event: CaptureEvent) {}
+        func onUpload(event: UploadEvent) {}
+        func onError(error: JoyfillError) {}
+        
+        func reset() {
+            capturedFocusEvents.removeAll()
+            capturedBlurEvents.removeAll()
+        }
```
Comment (VishnuBishnoi) [inline JoyfillSwiftUIExample/JoyfillTests/NavigationJSONTests.swift]: Nice helper. Since `ChangeFocusCapture` is a general-purpose `FormChangeEvent` mock for capturing focus/blur events, consider extracting it to a shared test utilities file — it could be useful for other test classes that need to verify focus/blur behavior (e.g., decorator action tests in the future).

File: JoyfillSwiftUIExample/JoyfillUITests/ChangeUITests/OnChangeHandlerUITests.swift [hunk 4]
```diff
@@ -1148,7 +1148,8 @@ final class OnChangeHandlerUITests: JoyfillUITestsBaseClass {
         }
         goToTableDetailPage()
         goToTableDetailPage(index: 0)
-        
+        firstScrollLeft()
```
Comment (VishnuBishnoi) [inline JoyfillSwiftUIExample/JoyfillUITests/ChangeUITests/OnChangeHandlerUITests.swift]: The added `firstScrollLeft()` / `secScrollLeft()` / `firstScrollRight()` calls appear to be compensating for a layout change (possibly the row decorator column from PR #292). If these tests are running against `main` (without decorators), these scrolls may be unnecessary or even break the tests.

Worth verifying that this PR's branch includes the decorator changes, or adding a comment explaining why the extra scrolling is needed.

**Review by VishnuBishnoi — COMMENTED** [review]
Good test coverage for the goto/navigation feature — the unit tests cover edge cases well (same page, hidden page, invalid page, round-trip), and the launch-arg approach for UI tests is a solid strategy to avoid flaky taps. A few concerns around test fragility below.

**Review by VishnuBishnoi — APPROVED** [review]
LGTM! The fragile test name routing was fixed with a static dictionary lookup, which was the main actionable concern. The rest of the design decisions (fixed delay, timeout values, helper structure) are reasonable for the test infrastructure being built here.
